spaghetto/src/Spago/Psa.purs

```diff
@@ -0,0 +1,205 @@
+import Spago.Purs as Purs
+
+type PsaArgs =
+  { libraryDirs :: Array String
+  , jsonErrors :: Boolean
+  , color :: Boolean
+  }
+
+defaultParseOptions :: PsaOptions
+defaultParseOptions =
+  { showSource: Core.ShowSourceCode
+  , stashFile: Nothing -- ".psa-stash"
+  , censorWarnings: false
+  , censorLib: false
+  , censorSrc: false
+  , censorCodes: Set.empty
+  , filterCodes: Set.empty
+  , statVerbosity: Core.CompactStats
+  , strict: false
+  }
+
+type PsaOptions =
+  { showSource :: Core.ShowSourceCode
+  , stashFile :: Maybe String
+  , censorWarnings :: Boolean
+  , censorLib :: Boolean
+  , censorSrc :: Boolean
+  , censorCodes :: Set ErrorCode
+  , filterCodes :: Set ErrorCode
+  , statVerbosity :: Core.StatVerbosity
+  , strict :: Boolean
+  }
+
+toOutputOptions :: PsaArgs -> PsaOptions -> PsaOutputOptions
+toOutputOptions { libraryDirs, color } options =
+  { color
+  , censorWarnings: options.censorWarnings
+  , censorLib: options.censorLib
+  , censorSrc: options.censorSrc
+  , censorCodes: options.censorCodes
+  , filterCodes: options.filterCodes
+  , statVerbosity: options.statVerbosity
+  , libraryDirs
+  , strict: options.strict
+  }
+
+psaCompile :: forall a. Set.Set FilePath -> Array String -> PsaArgs -> PsaOptions -> Spago (Purs.PursEnv a) Unit
+psaCompile globs pursArgs psaArgs options@{ showSource, stashFile } = do
+  let outputOptions = toOutputOptions psaArgs options
+  stashData <- case stashFile of
+    Just f -> readStashFile f
+    Nothing -> emptyStash
+
+  result <- Purs.compile globs (Array.snoc pursArgs "--json-errors")
+  let
+    result' = case result of
+      Left err -> { output: err.stdout, exitCode: err.exitCode, err: Just err }
+      Right success -> { output: success.stdout, exitCode: Just success.exitCode, err: Nothing }
+  arrErrorsIsEmpty <- forWithIndex (Str.split (Str.Pattern "\n") result'.output) \idx err ->
+    case jsonParser err >>= CA.decode psaResultCodec >>> lmap CA.printJsonDecodeError of
+      Left decodeErrMsg -> do
+        logDebug $ Array.intercalate "\n"
+          [ "Failed to decode PsaResult at index '" <> show idx <> "': " <> decodeErrMsg
+          , "Json was: " <> err
+          ]
+        -- Note to reviewer:
+        -- Should a stash decode error cause a non-zero exit?
+        pure false
+      Right out -> do
+        files <- liftEffect $ Ref.new FO.empty
+        let
+          loadLinesImpl = if showSource == Core.ShowSourceCode then loadLines files else loadNothing
+          filenames = insertFilenames (insertFilenames Set.empty out.errors) out.warnings
+        merged <- mergeWarnings filenames stashData.date stashData.stash out.warnings
+        for_ stashFile \f -> writeStashFile f merged
+
+        out' <- buildOutput loadLinesImpl outputOptions out { warnings = merged }
+
+        liftEffect $ if psaArgs.jsonErrors then JsonPrinter.print out' else DefaultPrinter.print outputOptions out'
+
+        pure $ FO.isEmpty out'.stats.allErrors
```

Comment:
The logic here might be wrong regarding `jsonErrors`. The current code will output the Psa-modified errors rather than what is directly received from `purs`. Should it? Otherwise, we could rewrite this to:
```purs
if psaArgs.jsonErrors then
   Console.log pursJsonErrors
   die ...
else ...
  DefaultPrinter.print ...
```
and drop the JsonPrinter completely.

Reply:
I am not sure how psa is changing the errors here?
But the thing is that editors might get confused with a stash file, so maybe we should output all the json errors anyways. @nwolverson likely has more useful input on this